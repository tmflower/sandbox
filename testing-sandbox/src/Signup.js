import { useState } from "react";
import axios from "axios";

export function Signup() {
    const [message, setMessage] = useState('');
    const [formData, setFormData] = useState({ username: '', password: ''});
    const { username, password } = formData;

    const handleChange = (evt) => {
        const { name, value } = evt.target;
        setFormData(formData => ({ ...formData, [name]: value}));
    }

    const handleSubmit = async (evt) => {
        evt.preventDefault();
        try {            
            await axios.post(`http://localhost:3001/signup`, { username, password });
            setMessage("Signup successful!");            
        }
        catch(err) {
            console.error(err);
        }
    }

    return (
        <div>
            <h1>Sign up</h1>
            <form>
                <label htmlFor="username">Username:
                <input 
                    type="text" 
                    name="username" 
                    value={username} 
                    id="username" 
                    placeholder="username"
                    onChange={handleChange}>
                </input>
                </label>
                <label htmlFor="password">Password:
                <input 
                    type="password"
                    name="password" 
                    value={password} 
                    id="password" 
                    placeholder="password"
                    onChange={handleChange}>
                </input>
                </label>
                <button onClick={handleSubmit}>Submit</button> 
            </form>
            {message ? <p data-testid="message">{message}</p> : null}
            
        </div>
    )
}