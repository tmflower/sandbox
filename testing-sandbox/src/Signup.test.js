import { render, screen } from "@testing-library/react";
import { Signup } from "./Signup";
import userEvent from "@testing-library/user-event";
import axios from "axios";

test("renders without crashing", function() {
    render(<Signup />);
});

test("matches snapshot", () => {
    const { asFragment } = render(<Signup />);
    expect(asFragment()).toMatchSnapshot();
  });

test("renders Signup component correctly", () => {
    render(<Signup />)
    const heading = screen.getByText("Sign up");
    const username = screen.getByPlaceholderText("username");
    const password = screen.getByPlaceholderText("password");
    expect(heading).toBeInTheDocument();
    expect(username).toBeInTheDocument();
    expect(password).toBeInTheDocument();
});

jest.mock("axios");

test("handles user signup", () => {
    const newUser = { username: "user", password: "abc123$"}
    axios.post.mockImplementation(() => Promise.resolve({ data: { newUser } }));
});

test("handles submit button click", async() => {
    render(<Signup />);
    const buttonElement = screen.getByRole("button");
    const message = screen.queryByText("Signup successful!");
    expect(message).not.toBeInTheDocument();

    await userEvent.click(buttonElement);

    const message2 = screen.queryByText("Signup successful!");
    expect(message2).toBeInTheDocument();
});