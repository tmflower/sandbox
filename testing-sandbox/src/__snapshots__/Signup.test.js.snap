// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div>
    <h1>
      Sign up
    </h1>
    <form>
      <label
        for="username"
      >
        Username:
        <input
          id="username"
          name="username"
          placeholder="username"
          type="text"
          value=""
        />
      </label>
      <label
        for="password"
      >
        Password:
        <input
          id="password"
          name="password"
          placeholder="password"
          type="password"
          value=""
        />
      </label>
      <button>
        Submit
      </button>
    </form>
  </div>
</DocumentFragment>
`;
